### DISCUSSION TITLE
Proof Key for Code Exchange by OAuth Public Clients - Implementation - React Native - Expo - SPA - OAuth2

### DISCUSSION THREAD

**Discussion by NilsBaumgartner1994** [directus/directus#23611] [Feature Requests]
### Summary

Implementation / Adoption of RFC 7636 (Proof Key for Code Exchange by OAuth Public Clients)

This feature request proposes integrating native support for PKCE (Proof Key for Code Exchange) in Directus to enhance OAuth2 flows, particularly for native and mobile applications. PKCE provides an additional layer of security by ensuring that the entity that initiated the OAuth2 flow is the same one completing it, preventing interception and code injection attacks.

Implementing an abstraction layer for PKCE for directus: https://www.rfc-editor.org/rfc/rfc7636

https://auth0.com/docs/get-started/authentication-and-authorization-flow/authorization-code-flow-with-pkce 

### Basic Example

Native App or SPA will create:
- ```code_verifier``` a random secret
- ```code_challenge``` the hash of the secret

The native app or SPA will call
POST: ```directus.com/proof-key-code-exchange/authorize```
Body: 
- ```provider``` (e.g. google)
- ```code_challenge``` (a hash of a private secret)
- ```code_challenge_method``` (e.g. S256)
- ```redirect_url```

Then directus checks if provider is set, the redirect is allowed. 

Now the directus magic happens and the user is redirected to the provider and the normal Auth Flow (OAuth2) is performed.
Directus redirect the client to a ```/save-session``` endpoint, saves the ```directus_refresh_token``` and finally redirect the user to the desired redirect. The native app or SPA application will get the ```authorization_code``` from the redirect url. 

The native app or SPA will now ask for the ```directus_refresh_token```.
POST: ```directus.com/proof-key-code-exchange/token```
Body:
- ```authorization_code``` the "place" where the server parked our ```code_challenge```
- ```code_verifier``` our proof, that we created the ```code_challenge```
The endpoint now checks if we have created the ```code_challenge```
```hash(code_verifier) === code_challenge```
If that is true, the backend knows we created the challenge, and can return us the saved ```directus_refresh_token```

The native app or SPA can now obtain the refresh token and use it as normal.

### Motivation

Why are we doing this? What use cases does it support? What is the expected outcome?

Native Apps cannot hold secrets in binary (could be decompiled) and have no unique scheme (there could be multiple apps on a device with the same scheme).

The motivation behind this request is to enhance security in OAuth2 flows, especially for native and mobile applications. The use of PKCE mitigates the risk of intercepted authorization codes being used maliciously, which is crucial in environments where the client secret cannot be securely stored, such as mobile apps.

Expected Outcome:

- Increased security for OAuth2 flows in Directus.
- Support for native and mobile applications without exposing client secrets.
- Compliance with OAuth2 best practices for public clients.

### Detailed Design

The implementation would involve:

    Extending the existing OAuth2 endpoints (/authorize and /token) in Directus to accept PKCE parameters (code_challenge, code_challenge_method, code_verifier).
    Adding server-side logic to validate the code_verifier against the stored code_challenge when exchanging the authorization code for an access token.
    Storing PKCE-related data securely (e.g., in Redis) to ensure temporary access and prevent replay attacks.
    Providing configuration options in Directus to enforce PKCE for specific OAuth providers.

Corner Cases:

    Handling cases where code_challenge_method is unsupported or missing.
    Ensuring that the flow handles expired or invalid codes gracefully with proper error messaging.
    Compatibility with existing OAuth2 configurations and avoiding breaking changes for existing Directus instances.

Edit: Here is my current implementation: https://github.com/rocket-meals/rocket-meals/blob/master/apps/backend/Backend/directusExtensions/directus-extension-rocket-meals-bundle/src/proof-key-code-exchange-endpoint/index.ts

please report any bugs or security issues.

**Comment by aidenfoxx** [comment]
Just for any Directus devs looking into this, OAuth and OpenID already uses PKCE under the hood in Directus. https://github.com/directus/directus/blob/main/api/src/auth/drivers/oauth2.ts#L146

**Reply by NilsBaumgartner1994** [reply]
Just wanted to clarify that, i cannot use it directly?
Not that i reinvented the wheel, but i cannot just directus call the PKCE with my current directus configuration?

**Reply by aidenfoxx** [reply]
Exactly. It's just under the hood and currently you can only authenticate through the URL + redirect approach I'm sure you're familiar with.

**Reply by NilsBaumgartner1994** [reply]
Thanks for clarification. Yes :-) sorry if i bothered and "spammed" with the redirect approach in the last time. But sadly there is currently no reliable way for native apps or SPA with directus and SSO.

**Reply by aidenfoxx** [reply]
I understand, and I think it's a good proposal to allow more flexibility. 😄

**Reply by NilsBaumgartner1994** [reply]
I hope too that it would boost directus more :-) <3 
I am facing the problem that i got the directus_session_token but i am unable to get a refresh token out of it.... why?

**Reply by aidenfoxx** [reply]
That I'm not sure of. I know the team refactored how tokens and cookies are returned, so it could be related to that. But I'm very out of the loop with Directus these days.

**Comment by kopino4** [comment]
Did i understand correctly this implementation could be used to make SSO in mobile apps work without the need of a webview experience? For example Flutter. Where the app could use the native SSO components?

If so, could you provide an idea how this implementation could be used with the https://pub.dev/packages/google_sign_in and https://pub.dev/packages/sign_in_with_apple SSO packages?

**Reply by ymys** [reply]
Not clearly yet. Directus doesn't provide any endpoints to exchange google token to directus token. I'm stuck. Maybe switch to other platform will solve my case.

**Reply by NilsBaumgartner1994** [reply]
> Not clearly yet. Directus doesn't provide any endpoints to exchange google token to directus token. I'm stuck. Maybe switch to other platform will solve my case.

I updated my post and linked my solution. Feel free to use or to help improving it.

**Comment by NilsBaumgartner1994** [comment]
That is correct. You will just need that Directus Supports PKCM Flow, which
is in top of oAuth. The App required to have an App-scheme (like a url
which opens the App directly on the phone).

Martin Kop ***@***.***> schrieb am Mi. 23. Apr. 2025 um 12:38:

**Reply by kopino4** [reply]
What does it mean Directus supporting PKCM Flow? Only integrating your implementation above? Or are there some additional requirements?

Then for clarification. I can get those attributes with the https://pub.dev/packages/google_sign_in package:
![image](https://github.com/user-attachments/assets/e3a6fc5f-24e2-4fea-8786-5b7c8a34c07e)

How would i then proceed to call those endpoints you provided?

**Comment by fahami** [comment]
I have followed your steps and implementing the custom endpoint from your PKCE implementation above in order to enable authentication flow using Google sign in from Flutter apps. The custom endpoint already works until the `The native app or SPA can now obtain the refresh token` step. I have received the `directus_session_token` and `directus_refresh_token`, but when I want to do refresh token with payload of `directus_refresh_token` and having header `Authorization: Bearer $directus_session_token` isn't working. What I am missed @NilsBaumgartner1994 ?

**Reply by fahami** [reply]
My bad, i forgot to set the `REFRESH_TOKEN_TTL`, after setting that variable the pkce flow is running smoothly

**Comment by NilsBaumgartner1994** [comment]
Nice to hear. Think the PKCE flow should be integrated into Directus core.

Mochammad Fahmi Ali ***@***.***> schrieb am Sa. 31. Mai 2025
um 10:21:

**Comment by armanhadifatah** [comment]
Is there any update on this , its very helpfull if we have directus Auth natively supporting OAuth like google apple and so on that way we can use directus in our Expo react native app

**Comment by pavkoccino** [comment]
@fahami @NilsBaumgartner1994 Could you provide guidance on (if possible) how to utilize this extension without the need of opening browser? Our issue is apple is blocking our update, because we open the `/login` part in the browser. I've managed to make the extension work, however I don't understand how to open the redirectUrl that I get from `/authorize` function natively so the OAuth flow can progress inside the app (via native system dialog, instead of the browser).

Is this something you've dealt with?

**Reply by NilsBaumgartner1994** [reply]
Sorry, Apple Blocked that the app opens safari (and leaves your app)?

**Reply by pavkoccino** [reply]
They have not blocked it in a sense of deleting the current version from the store, but for about month now, they are preventing us from update, because they do not like that the sign in is not happening natively inside the app.

Is there a way to do that or there is no way around opening the browser?

**Reply by NilsBaumgartner1994** [reply]
> They have not blocked it in a sense of deleting the current version from the store, but for about month now, they are preventing us from update, because they do not like that the sign in is not happening natively inside the app.
> 
> Is there a way to do that or there is no way around opening the browser?

Retry submitting. Apple review is sometimes tricky. In the guidelines https://developer.apple.com/app-store/review/guidelines/#login-services the provider requirements are mentioned.

I assume, you haven’t supported an Apple login option?

ask for clarification which requirement was not met.

**Comment by NilsBaumgartner1994** [comment]
@br41nslug is there some chance of getting this endpoint: https://github.com/rocket-meals/rocket-meals/blob/master/apps/backend/Backend/directusExtensions/directus-extension-rocket-meals-bundle/src/proof-key-code-exchange-endpoint/index.ts
Being integrated into Directus core? Are there any other security issues related to this? Should I create a pull request for this feature?
